src/measure_theory/pi_system.lean

```diff
@@ -0,0 +1,166 @@
+/-
+Copyright (c) 2021 Martin Zinkevich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Martin Zinkevich
+-/
+import measure_theory.measure_space
+import tactic.fin_cases
+
+/-!
+# Lemmas regarding `is_pi_system`.
+
+`is_pi_system` is similar, but not identical to, the classic π-system encountered in measure
+theory. In particular, it is not required to be nonempty, and it isn't closed under disjoint
+intersection (thus neither more nor less general than a typical π-system).
+
+## Main statements
+
+* `generate_pi_system g` gives the minimal pi system containing `g`.
+  This can be considered a Galois insertion into both measurable spaces and sets.
+
+* `generate_from_generate_pi_system_eq` proves that if you generate a pi_system
+  then a measurable space versus generating a measurable space gives the same result. This
+  is useful because there are connections between a independent sets that are pi systems
+  the generated independent spaces.
+
+* `mem_generate_pi_system_Union_elim` and `mem_generate_pi_system_Union_elim'` are theorems
+  that show that any element of the supremum of the union of a set of pi systems can be
+  represented as the intersection of a finite number of elements from these sets.
+
+## Implementation details
+
+* is_pi_system is a predicate, not a type. Thus, we don't explicitly define the galois
+  insertion, nor do we define a complete lattice. In theory, we could define a complete
+  lattice and galois insertion on subtype is_pi_system.
+
+-/
+
+open measure_theory measurable_space
+open_locale classical
+
+lemma is_pi_system.singleton {α} (S : set α) : is_pi_system ({S} : set (set α)) :=
+begin
+  intros s t h_s h_t h_ne,
+  rw [set.mem_singleton_iff.1 h_s, set.mem_singleton_iff.1 h_t, set.inter_self,
+      set.mem_singleton_iff],
+end
+
+inductive generate_pi_system {α} (g : set (set α)) : set (set α)
+| base {s : set α} (h_s : s ∈ g) : generate_pi_system s
+| inter {s t : set α} (h_s : generate_pi_system s)  (h_t : generate_pi_system t)
+  (h_nonempty : (s ∩ t).nonempty) : generate_pi_system (s ∩ t)
+
+lemma is_pi_system_generate_pi_system {α} (g : set (set α)) :
+  is_pi_system (generate_pi_system g) :=
+  λ s t h_s h_t h_nonempty, generate_pi_system.inter h_s h_t h_nonempty
+
+lemma generate_pi_system_subset {α} {g t : set (set α)} (h_t : is_pi_system t)
+  (h_sub : g ⊆ t) : generate_pi_system g ⊆ t :=
+ begin
```

Comment:
```suggestion
begin
```

Reply:
This was marked resolved, but the stray space is still there.